src/hotspot/share/opto/node.cpp

```diff
@@ -665,13 +646,35 @@ void Node::destruct(PhaseValues* phase) {
   }
   BarrierSetC2* bs = BarrierSet::barrier_set()->barrier_set_c2();
   bs->unregister_potential_barrier_node(this);
+
 #ifdef ASSERT
   // We will not actually delete the storage, but we'll make the node unusable.
   *(address*)this = badAddress;  // smash the C++ vtbl, probably
   _in = _out = (Node**) badAddress;
   _max = _cnt = _outmax = _outcnt = 0;
   compile->remove_modified_node(this);
```

Comment:
`remove_modified_node(this)` should be called before we trash node's fields.

Reply:
Shouldn't remove_modified_node always be called, regardless of build type? Having it in an ASSERT block is wierd.

Reply:
Also moved it up to just before smashing the vtbl.